telemetry/data_slow_query.go

```
@@ -0,0 +1,181 @@
+	// update CurrentSQBInfo first, then gen delta
+	if err := UpdateCurrentSQB(ctx); err != nil {
+		return nil, err
+	}
+	delta := CalculateDeltaSQB()
+	return delta, nil
+}
+
+// UpdateCurrentSQB records current slow query buckets
+func UpdateCurrentSQB(ctx sessionctx.Context) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			switch x := r.(type) {
+			case string:
+				err = errors.New(x)
+			case error:
+				err = x
+			default:
+				err = errors.New("unknown failure")
+			}
+		}
+	}()
+
+	value, err := querySlowQueryMetric(ctx) //TODO: judge error here
+	if err != nil {
+		logutil.BgLogger().Info("querySlowQueryMetric got error")
+		return err
+	}
+
+	if value.Type() != pmodel.ValVector {
+		return errors.New("Prom vector expected, got " + value.Type().String())
+	}
+	promVec := value.(pmodel.Vector)
+	for _, sample := range promVec {
+		metric := sample.Metric
+		bucketName := metric["le"] //hardcode bucket upper bound
+		CurrentSQBInfo[string(bucketName)] = int(sample.Value)
+	}
+	return nil
+}
+
+func querySlowQueryMetric(sctx sessionctx.Context) (result pmodel.Value, err error) {
+	// Add retry to avoid network error.
+	var prometheusAddr string
+	for i := 0; i < 5; i++ {
+		//TODO: the prometheus will be Integrated into the PD, then we need to query the prometheus in PD directly, which need change the quire API
+		prometheusAddr, err = infosync.GetPrometheusAddr()
+		if err == nil || err == infosync.ErrPrometheusAddrIsNotSet {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if err != nil {
+		return nil, err
+	}
+	promClient, err := api.NewClient(api.Config{
+		Address: prometheusAddr,
+	})
+	if err != nil {
+		return nil, err
+	}
+	promQLAPI := promv1.NewAPI(promClient)
+	promQL := "tidb_server_slow_query_process_duration_seconds_bucket{sql_type=\"general\"}"
+
+	ts := time.Now()
+	// Add retry to avoid network error.
+	ctx := context.TODO() // just use default context
```

do we need set timeout to fail fast?

---

Accepted.